src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/CommentHelper.java

```diff
@@ -500,31 +500,36 @@ public ReferenceTree getType(DocTree dtree) {
     }
 
     public DocTreePath getDocTreePath(DocTree dtree) {
-        if (dcTree == null && element instanceof ExecutableElement ee) {
-            return getInheritedDocTreePath(dtree, ee);
+        if (dcTree == null) {
+            // Element does not have a doc comment.
+            return getInheritedDocTreePath(dtree);
         }
-        if (path == null || dcTree == null || dtree == null) {
+        if (path == null || dtree == null) {
             return null;
         }
         DocTreePath dtPath = DocTreePath.getPath(path, dcTree, dtree);
-        if (dtPath == null && element instanceof ExecutableElement ee) {
-            // The overriding element has a doc tree, but it doesn't contain what we're looking for.
-            return getInheritedDocTreePath(dtree, ee);
-        }
-        return dtPath;
+        // Doc tree isn't in current element's comment, it must be inherited.
+        return dtPath == null ? getInheritedDocTreePath(dtree) : dtPath;
     }
 
-    private DocTreePath getInheritedDocTreePath(DocTree dtree, ExecutableElement ee) {
+    private DocTreePath getInheritedDocTreePath(DocTree dtree) {
         Utils utils = configuration.utils;
-        var docFinder = utils.docFinder();
-        Optional<ExecutableElement> inheritedDoc = docFinder.search(ee,
-                (m -> {
-                    Optional<ExecutableElement> optional = utils.getFullBody(m).isEmpty() ? Optional.empty() : Optional.of(m);
-                    return Result.fromOptional(optional);
-                })).toOptional();
-        return inheritedDoc.isEmpty() || inheritedDoc.get().equals(ee)
-                ? null
-                : utils.getCommentHelper(inheritedDoc.get()).getDocTreePath(dtree);
+        if (element instanceof ExecutableElement ee) {
```

Comment:
Technically, it is an official anti-pattern to use `instanceof` on Language Model classes. Yes, it works OK for the JDK implementation of `javax.lang.model`, but it is not necessarily true for all other implementations (with a known instance where it is not the case.)

The recommended solution is to either use visitors or test or switch on the `ElementKind`.

That all being said, I accept this is in JDK code ... and now that we have pattern-switch available ...

Reply:
On the one hand, I'd like to suggest using pattern-switch -- on the other hand, I note that makes back porting harder.